flowmachine/flowmachine/features/subscriber/home_location_monthly.py

```diff
@@ -0,0 +1,198 @@
+            )
+            for day in rrule(DAILY, dtstart=full_range_start, until=full_range_stop)
+        }
+        modal_days = [
+            query
+            for (day, query) in all_locations.items()
+            if full_range_start <= day <= full_range_stop
+        ]
+        self.last_locations = [
+            query
+            for (day, query) in all_locations.items()
+            if self._window_start <= day <= self._window_stop
+        ]
+        self.modal_locations = ModalLocation(*modal_days)
+
+        # self.daily_location_frequency = LocationVisits(
+        #     self.last_locations  # Ask Jono about this
+        # )
+
+        super().__init__()
+
+    @property
+    def column_names(self) -> List[str]:
+        cols = ["subscriber", "location", "month"]
+        return cols
+
+    @property
+    def window_start(self):
+        return self._window_start.strftime("%Y-%m-%d")
+
+    @window_start.setter
+    def window_start(self, value):
+        if type(value) is str:
+            self._window_start = datetime.strptime(value, "%Y-%m-%d")
+        elif type(value) is datetime:
+            self._window_start = value
+        else:
+            raise TypeError("window_start must be datetime or yyyy-mm-dd")
+
+    @property
+    def window_stop(self):
+        return self._window_stop.strftime("%Y-%m-%d")
+
+    @window_stop.setter
+    def window_stop(self, value):
+        if type(value) is str:
+            self._window_stop = datetime.strptime(value, "%Y-%m-%d")
+        elif type(value) is datetime:
+            self._window_stop = value
+        else:
+            raise TypeError("window_stop must be datetime or yyyy-mm-dd")
+
+    def _make_query(self):
+
+        last_locations_clause = ""
+
+        # NOTE: last_locations seems to be dropping the final record? This is weird.
+        for last_location in self.last_locations:
+            last_locations_clause += f"""
+SELECT subscriber, pcod, '{last_location.start}' AS day
+FROM ({last_location.get_query()}) AS tbl
+UNION ALL
+"""
+
+        last_locations_clause = last_locations_clause.rstrip("UNION ALL\n")
+
+        sql = f"""
+WITH last_locations AS (
+    {last_locations_clause}
+), modal_locations AS (
+    {self.modal_locations.get_query()}
+), location_histogram AS (
+	SELECT subscriber, pcod, count(pcod) as y
+	FROM last_locations
+	GROUP BY subscriber, pcod
+), this_month_known_homes AS (
+	SELECT modal_locations.subscriber, modal_locations.pcod 
+	FROM modal_locations
+		INNER JOIN  location_histogram ON location_histogram.subscriber = modal_locations.subscriber
+	WHERE modal_locations.pcod = location_histogram.pcod
+	AND modal_locations.subscriber = location_histogram.subscriber
+	AND location_histogram.y >= {self.home_this_month}
+), """
+
+        if self.ref_location:
+            sql += f"""
+reference_locations AS (
+    SELECT subscriber, pcod
+    FROM ({self.ref_location.get_query()}) AS tbl
+), last_month_known_homes AS (
```

Comment:
Personally, I'd be tempted to combine 'this_month_known_homes' and 'last_month_known_homes', on the basis that the query result should have one row for each row in 'modal_locations' so it feels neater to avoid the `subscriber NOT IN (SELECT ...)`, and only join 'modal_locations' to 'location_histogram' once.

Something like this (square brackets mark reference_location-dependent lines):
```
  SELECT modal_locations.subscriber, modal_locations.pcod
  FROM modal_locations
  INNER JOIN location_histogram USING ( subscriber, pcod )
[ LEFT JOIN reference_locations USING (subscriber)  ]
  WHERE location_histogram.y >= {self.home_this_month}
[ OR ( modal_locations.pcod = reference_locations.pcod AND location_histogram.y >= {self.home_last_month} )  ]
```

It's a matter of preference though, I suppose - what seems neater to me might seem less so to others.

Reply:
Could even combine 'unknown_homes' into this as well, by moving the where condition into a `CASE WHEN` in the select:
```
  SELECT modal_locations.subscriber,
         CASE WHEN ( location_histogram.y >= {self.home_this_month} )
[             OR ( modal_locations.pcod = reference_locations.pcod AND location_histogram.y >= {self.home_last_month} )  ]
              THEN modal_locations.pcod
              ELSE 'unknown'
         END AS pcod
  FROM ...
```

But again, personal preference...